Add support for AuthorizedKeysCommand and AuthorizedPrincipalsCommand to run as System

bkatyl opened the pull request:
Add support for AuthorizedKeysCommand and AuthorizedPrincipalsCommand to run as System.

- When sshd is run as system and provided username is system, run command as subprocess using posix_spawnp
- Improve debug logging in spawn_child_internal by providing different logging message for when CreateProcessW and CreateProcessAsUserW are called.

This is part 2 of the a 2 part PR to Fix issue #1546

bkatyl committed:
Add support for AuthorizedKeysCommand and AuthorizedPrincipalsCommand to run as System

Add support for AuthorizedKeysCommand and AuthorizedPrincipalsCommand to run as System.

- When sshd is run as system and provided username is system, run command as subprocess using posix_spawnp
- Improve debug logging in spawn_child_internal by providing different logging message for when CreateProcessW and CreateProcessAsUserW are called.

This is part 2 of the a 2 part PR to Fix issue #1546

bagajjal commented on auth.c:
```diff
@@ -955,8 +955,18 @@ subprocess(const char *tag, struct passwd *pw, const char *command,
-		else if (__posix_spawn_asuser((pid_t*)&pid, av[0], &actions, NULL, av, NULL, pw->pw_name) != 0)
-			fatal("posix_spawn: %s", strerror(errno));
+		else {
+			if (strcmp(pw->pw_name, "system") == 0 && am_system()) {
```
@bkatyl - This mayn't succeed on a windows machine with non-english culture. It's better to get the SID and compare the SID with WinLocalSystemSid. 
Please refer to the logic [here](https://github.com/PowerShell/openssh-portable/blob/1a1a2ac5f1720f791416f3f16e15b369a5f05fe9/contrib/win32/win32compat/w32-sshfileperm.c#L122)

bkatyl replied:
Thanks for the feedback and pointer to the logic, it was very helpful. 

I added a commit that includes the changes to SID comparison. Since these are windows only calls, I moved the evaluation into the Windows ifdef. The needed imports are also guarded by a Windows ifdef.

bagajjal commented on contrib/win32/win32compat/w32fd.c:
```diff
@@ -1074,11 +1074,14 @@ spawn_child_internal(const char* cmd, char *const argv[], HANDLE in, HANDLE out,
-		debug3("spawning %ls", t);
-		if (as_user)
+		if (as_user) {
+			debug3("spawning %ls as user", t);
```
Thanks for fixing this.

bagajjal commented:
Please resolve the conflicts in the auth.c file.

bkatyl committed:
Updating to compare SID with WinLocalSystemSid

- Added the needed includes, guarded by Windows ifdef.
- Added boolean variable to store determination with default of false.
- Moved evaluation into Windows only section since we're using SID comparison.
- Changed from string evaluation of "system" to comparing the SID of the user with WinLocalSystemSid.

bkatyl committed:
Merge branch 'latestw_all' into subprocess

bagajjal commented on auth.c:
```diff
@@ -934,6 +939,10 @@ subprocess(const char *tag, struct passwd *pw, const char *command,
 		restore_uid();
 		return 0;
 	}
+	/* If the user's SID is the System SID and sshd is running as
```
Can you please move this and line 902 to line 971.
Just want to have the code near where it's used and also this whole thing is windows specifc.

bkatyl replied:
Absolutely. 
I took "whole thing is windows specifc" to mean that the use of IsWellKnownSid didn't need to be in a "#ifdef WINDOWS", so I collapsed it all into the if statement. If I read that incorrectly, please let me know as I'm also happy to had the indef and variable to that specific section.

bagajjal replied:
Thank you for fixing it. 
Could you please fix the comments in your other PR (https://github.com/PowerShell/openssh-portable/pull/478).
I would like to verify the changes before merging. Please provide the steps to verify.

bkatyl committed:
Consolidating changes to be in one part of the file

Consolidating changes to be in one part of the file
